### ISSUE TITLE
Puppeteer - I.getCurrentUrl() returns last URL rather than current url

### ISSUE THREAD

**Issue by awshanks** [Codeception/CodeceptJS#963]
#### What do you get instead?

> Provide console output if related. Use `--verbose` mode for more details.

```bash
Open Github @sample --
 open browser and navigate to github.com
 • I am on page "https://github.com/awshanks"
 • I click "codecept_training_material"
 • I get current url 
https://github.com/awshanks
```

> Provide test source code if related

```js
Scenario('open browser and navigate to github.com', (I) => {
I.amOnPage('https://github.com/awshanks');
I.click('codecept_training_material');
I.getCurrentUrl().then(function (result) {
    console.log(result);
});
});
```

### Details

* CodeceptJS version: 1.1.5 (puppeteer 1.1.0)
* NodeJS Version: 8.9.4
* Operating System: MacOS
* Protractor || WebDriverIO || Nightmare version (if related)
* Configuration file:

```js
# paste suite config here
```

**Comment by reubenmiller** [comment]
The `I.getCurrentUrl` Function does not exist on the Puppeteer helper. Are you using a custom helper? If so then it might be a bug with puppeteer and not codeceptjs. 

But could you please elaborate and fill out the issue template in full so we can address the problem more easily.

**Comment by awshanks** [comment]
Yes, I just added the function to the helper in the codecept node_module folder (saw the pull request for it) (as puppeteer custom helpers do not work right now).

I added the code above.

From https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteerlaunchoptions -
 Bare in mind that if click() triggers a navigation event and there's a separate page.waitForNavigation() promise to be resolved, you may end up with a race condition that yields unexpected results. The correct pattern for click and wait for navigation is the following:
`
const [response] = await Promise.all([
  page.waitForNavigation(waitOptions),
  page.click(selector, clickOptions),
]);
`

**Comment by reubenmiller** [comment]
@awshanks There are few things that are not clear to me.

1. What do you mean that custom Puppeteer helpers do not work? You should be able to create a custom helper for any of the existing helpers. If you have a problem with the config then maybe I can help with that.

1. For clicks that cause a navigation action, you need to wait for the navigation manually. There is a comment about this in #914. We can't use the `const [response] = await Promise.all([ page.waitForNavigation(waitOptions), page.click(selector, clickOptions), ]);` by default because it will not work on elements that don't trigger a navigation action. However you could use the new retry feature introduced by @DavertMik in v1.1.5, so your test would look like:

```js
Scenario('open browser and navigate to github.com', (I) => {
  I.amOnPage('https://github.com/awshanks');
  I.click('codecept_training_material');
  I.retry(3).seeInCurrentUrl('codecept_training_material');
});
```

**Comment by awshanks** [comment]
Okay, thanks for the workaround.  Glad to have it documented here for reference.

**Comment by awshanks** [comment]
For the Puppeteer custom helper, I have the following function (works for selenium, nightmare etc, only addition is the **page**)

`
getCurrentURL() {
        var browser = this.helpers['Puppeteer'].browser;
        return browser.page.url();
    }
`
Is this not correct?

**Comment by reubenmiller** [comment]
No the [Puppeteer api](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md) is a little different to WebDriverIO etc.

The `.url()` function is actually on the **Page** object and not the **Browser**. Though in the helper we keep track of the current page in the `.page` property.

So an example custom Helper function would look like this.

```js
async getCurrentUrl() {
  const helper = this.helpers['Puppeteer'];
  return helper.page.url();
}
```

**Comment by reubenmiller** [comment]
Here is the full example of a custom helper for sake of completeness.

```js
'use strict';

let Helper = codecept_helper;

class MyCustomHelper extends Helper {
  async getCurrentUrl() {
    const helper = this.helpers['Puppeteer'];
    return helper.page.url();
  }
}

module.exports = MyCustomHelper;
```

**Comment by awshanks** [comment]
Ah, I see, maybe an example in the puppeteer documentation would help others?

Again thanks for clarifying.

**Comment by reubenmiller** [comment]
@awshanks Not a problem I have just submitted a PR with a comment about how to access it from a custom Helper (like the example in the WebDriverIO). It will take a little bit before it is available from the doc website.

**Comment by awshanks** [comment]
Okay, I think I've narrowed down the issue I was having and it seems that if a website use hash routing the page.url doesn't update after a click that directs to another page.

The I.click works fine and goes to the page and I see the url in the verbose output.

If I go to the page with I.amOnPage('/#/page') the route will change but codecept doesn't know that the page has changed and the amOnPage will fails after it retries and I always just have the baseUrl that I defined in the config.

So this is possibly an issue with codecept / puppeteer.  The same steps worked in webdriver / nightmare before this.
